[LDM] Enforce LDM Font Policy in GPUP & avoid trusted list SFP is enabled

vitorroriz opened the pull request:
#### 5b8f348c9f76a60a96c3a6ce9cd171d609c4bc77
<pre>
[LDM] Enforce LDM Font Policy in GPUP &amp; avoid trusted list SFP is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=272247">https://bugs.webkit.org/show_bug.cgi?id=272247</a>
<a href="https://rdar.apple.com/124235570">rdar://124235570</a>

Reviewed by Brent Fulgham.

If SafeFontParser is enabled we should use the SFP on both WCP and GPUP sides
and we should not parse trusted fonts with the system font parser.

When SFP is disabled we should switch to using the system parser but only
for trusted fonts, like before.

The difference after this patch is that if the SafeFontParser is enabled
we will no longer check first for trusted fonts that can be parsed
with the system font parser like before.

Canonical link: <a href="https://commits.webkit.org/277717@main">https://commits.webkit.org/277717@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/afb353ecfb9410092f5a4279877863877a1e62d2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48366 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27578 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51320 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51054 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44431 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50671 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33514 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25099 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48948 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25263 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/41765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20651 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22743 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6422 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/44695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52959 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/23413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/19748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46844 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24678 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/41956 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/45759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10671 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/25483 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24401 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/loader/cache/CachedFont.h:
```diff
@@ -89,7 +89,7 @@ class CachedFont : public CachedResource {
 
     FontParsingPolicy m_fontParsingPolicy { FontParsingPolicy::Deny };
-    bool m_didRefuseToParseCustomFont { false };
+    bool m_didRefuseToParseCustomFontWithSafeFontParser { false };
```
It doesn't seem important to change the name of this variable if all parsing switches to the new 'Safe Font Parser' in this patch. Just leave the name alone.

brentfulgham commented on Source/WebCore/loader/cache/CachedFont.h:
```diff
@@ -61,7 +61,7 @@ class CachedFont : public CachedResource {
     virtual RefPtr<Font> createFont(const FontDescription&, bool syntheticBold, bool syntheticItalic, const FontCreationContext&);
 
-    bool didRefuseToParseCustomFont() const { return m_didRefuseToParseCustomFont; }
+    bool didRefuseToParseCustomFontWithSafeFontParser() const { return m_didRefuseToParseCustomFontWithSafeFontParser; }
```
It doesn't seem important to change the name of this variable if all parsing switches to the new 'Safe Font Parser' in this patch. Just leave the name alone.

brentfulgham commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -369,9 +372,7 @@ void RemoteRenderingBackend::cacheFontCustomPlatformData(WebCore::FontCustomPlat
-    // FIXME: (rdar://124235570) use this->shoulUsedLockdownFontParser instead of hard-coded 'false' at tryMakeFromSerializationData after we deprecate lockdown mode fonts allowed (trusted) list.
-    constexpr bool shouldUseLockdownFontParser { false };
-    auto customPlatformData = FontCustomPlatformData::tryMakeFromSerializationData(WTFMove(fontCustomPlatformSerializedData), shouldUseLockdownFontParser);
+    auto customPlatformData = FontCustomPlatformData::tryMakeFromSerializationData(WTFMove(fontCustomPlatformSerializedData), this->shouldUseLockdownFontParser());
```
Nit: I don't think `this` is necessary anymore, now that the overloaded `shouldUseLockdownFontParser` local has been removed.

brentfulgham commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -612,10 +613,12 @@ void RemoteRenderingBackend::terminateWebProcess(ASCIILiteral message)
 bool RemoteRenderingBackend::shouldUseLockdownFontParser() const
 {
-    return m_gpuConnectionToWebProcess->isLockdownSafeFontParserEnabled() && m_gpuConnectionToWebProcess->isLockdownModeEnabled();
+    return m_gpuConnectionToWebProcess->isLockdownSafeFontParserEnabled() && m_gpuConnectionToWebProcess->isLockdownModeEnabled() && PAL::canLoad_CoreText_CTFontManagerCreateMemorySafeFontDescriptorFromData();
```
Nit: `PAL::canLoad_CoreText_CTFontManagerCreateMemorySafeFontDescriptorFromData()` probably doesn't change after launch. Maybe this could be captured at launch time and remembered for later. (Not necessary for this patch).

brentfulgham reviewed (approved):
Looks good -- I suggest not renaming since we aren't distinguishing between "Safe Font Parser" and some other font parser -- there's only one controlled by this flag -- so just keep the old naming. r=me

vitorroriz committed:
[LDM] Enforce LDM Font Policy in GPUP & avoid trusted list SFP is enabled
https://bugs.webkit.org/show_bug.cgi?id=272247
rdar://124235570

Reviewed by Brent Fulgham.

If SafeFontParser is enabled we should use the SFP on both WCP and GPUP sides
and we should not parse trusted fonts with the system font parser.

When SFP is disabled we should switch to using the system parser but only
for trusted fonts, like before.

The difference after this patch is that if the SafeFontParser is enabled
we will no longer check first for trusted fonts that can be parsed
with the system font parser like before.

Canonical link: https://commits.webkit.org/277717@main
